### ISSUE TITLE
Ninja builder: generated UFO instances do not match source UFOs

### ISSUE THREAD

**Issue by bramstein** [googlefonts/gftools#843]
I'm trying to build a family of static fonts (one UFO source per style), but when using the new Ninja-based builder I noticed a lot of the "custom" settings were not in the final fonts:

```diff
< postscript name: P22UndergroundPro-LightItalic
< unique id: 4.000;P22;P22UndergroundPro-LightItalic
---
> postscript name: P22Underground-LightItalic
> unique id: 4.000;P22;P22Underground-LightItalic
268c268
< fsSelection italic: True
---
> fsSelection italic: False
275c275
< macStyle italic: True
---
> macStyle italic: False
```

Most notably, it doesn't seem to set the italic bits correctly when generating the "instance" UFOs and doesn't copy the postscript name. There might be other changes as well, but these were most visible. This seems to happen in the step where it generates instance UFOs from the generated designspace files.

**Comment by simoncozens** [comment]
Thanks, that's super interesting. We generate those instance UFOs from the design space using fontmake -i -o UFOs; ie we're not doing anything unusual inside the builder here. So this may indicate a fontmake bug. I'll try to reproduce it myself, but any reproducers that can be made public will be helpful.

As a side note we are right on the point of merging the new builder2; if we can find out what's going wrong here, it will produce a useful test case to make sure it's fixed in builder2 as well

**Comment by simoncozens** [comment]
TLDR: Can you not list each UFO file under the `sources:` key in your config file?

---

OK, this is complicated. And it does happen in builder2, although it's potentially easier to fix there.

1) I took a copy of a UFO font (Noto Serif Tamil Italic, as it happens) and built a designspace around it: one master, one master with the same coordinates.
2) I then edited the UFO file and added some custom values to the `fontinfo.plist`:

```xml
    <key>postscriptFontName</key>
    <string>NotoSerifTamilPro-Italic</string>
```

3) I ran `fontmake -o ufo -i -m test.designspace`.
4) I inspected `instance_ufo/NotoSerifTamil-Italic.ufo/fontinfo.plist`, and the `postscriptFontName` is not there.

I believe this is the cause of what you're seeing. Another way to see that is that `fontmake -o ttf -i -m test.designspace` creates a font with nameID6=NotoSerifTamil-Italic but `fontmake -o ttf -m test.designspace` creates a font with nameID6=NotoSerifTamilPro-Italic.

I don't know if this is a fontmake bug or not. Throwing away information explicitly provided in the UFO seems bad, but on the other hand reconstructing the instance's PostScript name based on its position in the "designspace" (such as it is) is exactly what you would want an instance generator to do. So maybe fontmake is being good here and we're just doing something we shouldn't. I'm not sure.

Equally what you seem to be wanting to do is to build a font which is defined in the `<instance>` element of a designspace file without, as it were, running an instance instantiator on the designspace file - is that a reasonable thing to want to do? Again, I'm not sure.

Before the ninja builder, we could set the option `interpolate: false` (which was also set automatically to false when buildVariable was set to false), but we have dropped this and all UFOs which will eventually be built into binaries are shepherded through `fontmake -i -o ufo`.  This is made slightly harder to fix by the fact that `fontmake -o ufo -m test.designspace` does nothing - not unreasonably.

Is there a particular reason why you need to wrap these sources in a designspace file? Given that they're not interpolated, not instanced, and not variable, I'm not convinced a designspace actually adds anything to the equation here. You might find it easier just to list the UFO names under `sources:` instead.

**Comment by anthrotype** [comment]
when you run fontmake -i, it goes through it's fontmake.instantiator which among other things copies a bunch of static (non interpolatable) fontinfo fields from the default source to all the instance UFOs, see:

https://github.com/googlefonts/fontmake/blob/0971eb0ed3b242f449d36d210f89b5e8d537c3f6/Lib/fontmake/instantiator.py#L78-L148

Ideally, fontmake.instantiator would read and apply the [newly defined `public.fontInfo` lib key](https://github.com/fonttools/fonttools/pull/3358) that may be present in a designspace `<instance><lib>` element.

**Comment by anthrotype** [comment]
when building an interpolated instance, fontmake also runs glyphsLib.apply_instance_data which will apply any Glyphs.app's instance-specific custom parameters (that glyphsLib has temporarily placed in the designspace lib) to the generated instances. This allows to set things like postscriptFontName and similar.
If your workflow starts with a designspace and not with a .glyphs source, this should still work, but you'd have to put the data in the DS in the same way glyphsLib would...

**Comment by simoncozens** [comment]
OK, so you can use attributes on the designspace file to copy stuff into the instance - it looks like you can set the PostScript name using the `postscriptfontname` attribute on the instance, and fontmake (and therefore gftools-builder) will respect that; similarly macStyle should be set via `stylemapstylename`. Not sure about OS/2 selection bits though.

**Comment by anthrotype** [comment]
> OS/2 selection bits

if you're referring to italic/bold stuff, that's handled the same as head.macStyle using the `stylemapstylename` attribute of DS instance elements.

For other things, one could use Glyphs.app instance custom parameters stored in the DS instance lib, e.g. see below for fsType for example:

```xml
    <instance name="Test Regular" familyname="Test" stylename="Regular" filename="../instance_ufo/Test-Regular.ufo" stylemapfamilyname="Test Regular" stylemapstylename="bold">
      <location>
        <dimension name="Weight" xvalue="90"/>
      </location>
      <lib>
        <dict>
          <key>com.schriftgestaltung.customParameters</key>
          <array>
            <array>
              <string>fsType</string>
              <array>
                <integer>3</integer>
              </array>
            </array>
          </array>
        </dict>
      </lib>
    </instance>
```

Looks like "com.schriftgestaltung.customParameters" is an array of arrays (or list of tuples where first item is the parameter name (string) and the second whatever the parameter contains, in this particular case another list of numbers)

**Comment by bramstein** [comment]
Here is a minimal test case:

[FamilyTest.zip](https://github.com/googlefonts/gftools/files/14496805/FamilyTest.zip)

I run the following commands:

```sh
gftools builder FamilyTest/config.yaml
ttx FamilyTest/fonts/ttf/FamilyTest-RegularItalic.ttf

fontmake -u FamilyTest/sources/ufo/FamilyTest-RegularItalic.ufo
ttx master_ttf/FamilyTest-RegularItalic.ttf

diff master_ttf/FamilyTest-RegularItalic.ttx FamilyTest/fonts/ttf/FamilyTest-RegularItalic.ttx
```

The resulting diff (minus irrelevant changes):

```diff
     <xMax value="1024"/>
     <yMax value="1024"/>
-    <macStyle value="00000000 00000010"/>
+    <macStyle value="00000000 00000000"/>
     <lowestRecPPEM value="6"/>
     <fontDirectionHint value="2"/>
     <indexToLocFormat value="0"/>
@@ -74,7 +74,7 @@
     <xAvgCharWidth value="1024"/>
     <usWeightClass value="400"/>
     <usWidthClass value="5"/>
-    <fsType value="00000000 00000100"/>
+    <fsType value="00000000 00001000"/>
     <ySubscriptXSize value="0"/>
     <ySubscriptYSize value="0"/>
@@ -103,7 +103,7 @@
     <ulUnicodeRange3 value="00000000 00000000 00000000 00000000"/>
     <ulUnicodeRange4 value="00000000 00000000 00000000 00000000"/>
     <achVendID value="    "/>
-    <fsSelection value="00000000 00000001"/>
+    <fsSelection value="00000000 01000000"/>
     <usFirstCharIndex value="64"/>
     <usLastCharIndex value="64"/>
     <sTypoAscender value="1024"/>
@@ -135,6 +135,17 @@

@@ -159,13 +170,13 @@
 
   <name>
     <namerecord nameID="1" platformID="3" platEncID="1" langID="0x409">
-      Family Test
+      Family Test Regular Italic
     </namerecord>
     <namerecord nameID="2" platformID="3" platEncID="1" langID="0x409">
-      Italic
+      Regular
     </namerecord>
     <namerecord nameID="3" platformID="3" platEncID="1" langID="0x409">
-      1.000;;FamilyTestPro-RegularItalic
+      1.000;;FamilyTest-RegularItalic
     </namerecord>
     <namerecord nameID="4" platformID="3" platEncID="1" langID="0x409">
       Family Test Regular Italic
@@ -174,7 +185,7 @@
       Version 1.000
     </namerecord>
     <namerecord nameID="6" platformID="3" platEncID="1" langID="0x409">
-      FamilyTestPro-RegularItalic
+      FamilyTest-RegularItalic
     </namerecord>
     <namerecord nameID="16" platformID="3" platEncID="1" langID="0x409">
       Family Test
```

**Comment by simoncozens** [comment]
Ah, OK; thank you for the reproducer! This is fixed in builder2, where we don't try wrapping a raw UFO in a designspace.

@m4rc1e, another reason we should merge soon. ;-)

**Comment by bramstein** [comment]
@simoncozens Thanks, good to know it is fixed in builder2. Any rough timeline for merging and releasing it?

**Comment by anthrotype** [comment]
oh so the source in question is a single UFO and gftools was wrapping that in a designspace with a single master, single instance?

I still believe that fontmake should produce the same result whether you build the single UFO directly or via the wrapping designspace...
Maybe the wrapping DS was somehow missing some pieces of data

**Comment by bramstein** [comment]
@anthrotype Yes, that's what is happening. Before the Ninja builder building basically resulted in a `fontmake -u ...` call, but the designspace wrapping/instance generation seems to cause the issue. I agree that fontmake behaviour should be consistent.

**Comment by simoncozens** [comment]
Builder2 is now merged. Release later this week or early next.
